src/parser/core/modules/PrecastStatus.ts

```diff
@@ -0,0 +1,153 @@
+						// Ignore the applybuff event and check the applybuffstack event for validity
+						if (event.type === 'applybuff') { continue }
+						if (event.type === 'applybuffstack')
+						{
+							if (event.stack < statusInfo.stacksApplied) {
+								// First applybuffstack seen with less than max stacks, synth the initial buff event
+								this.fabricateBuffEvent(event, statusInfo)
+							}
+							else {
+								// First applybuff stack seen with max stacks, check if cast event needs to be synthesized
+								this.fabricateCastEvent(event, statusInfo)
+							}
+							this.markStatusAsTracked(event.ability.guid, event.targetID)
+						}
+					} else {
+						// This action does not apply stacks, check to see if this is an applybuff event
+						if (event.type === 'applybuff') {
+							// First event for this status is an apply buff, check if cast event needs to be synthesized
+							this.fabricateCastEvent(event, statusInfo)
+							this.markStatusAsTracked(event.ability.guid, event.targetID)
+						}
+					}
+
+					if (['removebuff', 'removebuffstack', 'refreshbuff'].includes(event.type)) {
+						this.fabricateBuffEvent(event, statusInfo)
+						this.markStatusAsTracked(event.ability.guid, event.targetID)
+					}
+				}
+			}
+		}
+
+		return [...this.castEventsToSynth, ...this.buffEventsToSynth, ...events]
+	}
+
+	private fabricateBuffEvent(event: AbilityEvent, statusInfo: Status) {
+		this.debug(`Fabricating applybuff event for status ${statusInfo.name} on target ${event.targetID}`)
+		// Fab an event and splice it in at the start of the fight
+		this.buffEventsToSynth.push({
+			...event,
+			timestamp: this.startTime - 1,
+			type: 'applybuff',
+		})
+
+		if (statusInfo.stacksApplied && statusInfo.stacksApplied > 0) {
+			this.debug(`Fabricating applybuff event for status ${statusInfo.name} with ${statusInfo.stacksApplied} stacks`)
+			// Status applies multiple stacks - fab an applybuffstack event
+			this.buffEventsToSynth.push({
+				// Can inherit most of the event data from the current one
+				...event,
+				// Override a few vals
+				timestamp: this.startTime - 1,
+				type: 'applybuffstack',
+				stack: statusInfo.stacksApplied,
+			})
+		}
+
+		// Determine if this buff comes from a known action, fab a cast event
+		this.fabricateCastEvent(event, statusInfo)
+	}
+
+	private fabricateCastEvent(event: AbilityEvent, statusInfo: Status) {
+		this.debug(`Determining if status ${statusInfo.name} was applied by a known action`)
+		// Determine if this buff comes from a known action, fab a cast event
+		const actionInfo = this.data.getActionAppliedByStatus(statusInfo)
+
+		if (!actionInfo) {
+			this.debug('No known action found, no cast event to synthesize')
+		} else {
+			if (this.trackedActions.includes(actionInfo.id)) {
+				this.debug(`Cast event already seen for ${actionInfo.name}, no synth necessary`)
+			} else {
+				this.debug(`Fabricating cast event for action ${actionInfo.name} by ${event.sourceID}`)
```

Comment:
this feels like a lot of extra mental overhead to read for the sake of some logging - while i'm on board with logging debug info, it shouldn't be at the cost of the code being readable.

if you've got any ideas on how to improve logging so this is less of an issue in-place i'm all ears

Reply:
I could at least improve the if/else nesting by switching to an "if/log/early return" pattern instead for checking the pre-conditions.  Personally, I really want this pre-condition debug logging to stay in place one way or another -- the hardest thing to trace in normalisers is figuring out why they are or are not choosing to normalise an event, so I consider this logging fairly high priority, even at the expense of "noop" if blocks.